src/Pdf/index.ts

```diff
@@ -0,0 +1,254 @@
+import { getRoutesForTripDay } from '../Route';
+import { Day, getTripDetailed, ItineraryItem, Trip } from '../Trip';
+import { getStaticMap } from './DataAccess';
+import { PdfData, PdfDestination, PdfQuery, StaticMap, StaticMapSector} from './PdfData';
+
+export async function getPdfData(query: PdfQuery): Promise<PdfData> {
+	const trip: Trip = await getTripDetailed(query.tripId);
+	const pdfData: PdfData = {
+		destinations: [],
+		routes: []
+	};
+
+	if (!trip.days) {
+		return pdfData;
+	}
+
+	pdfData.destinations = await extractDestinationsFromTrip(trip);
+	pdfData.destinations = fillDestinationsWithPlacesFromCollections(
+		pdfData.destinations,
+		await getCollectionsForDestinations(pdfData.destinations)
+	);
+
+	pdfData.destinations = await generateDestinationsMaps(pdfData.destinations, query);
+
+	pdfData.routes = await Promise.all(trip.days.map((day: Day, dayIndex: number) => (
+		getRoutesForTripDay(trip.id, dayIndex)
+	)));
+
+	return pdfData;
+}
+
+export async function extractDestinationsFromTrip(trip: Trip): Promise<PdfDestination[]> {
+	interface PdfDestinationDataTmp {
+		destinationId: string;
+		mainMap: StaticMap|null;
+		secondaryMaps: StaticMap[];
+		places: Place[];
+	}
+
+	const destinations: PdfDestinationDataTmp[] = [];
+
+	trip.days!.forEach((day: Day) => {
+		day.itinerary.forEach(async (itineraryItem: ItineraryItem) => {
+			const itineraryItemDestinationId: string = getFirstCity(itineraryItem.place!);
+			const destination: PdfDestinationDataTmp|undefined = destinations.find((d) => (
+				d.destinationId === itineraryItemDestinationId
+			));
+			if (destination) {
+				const place: Place|undefined = destination.places.find((p: Place) => p.id === itineraryItem.place!.id);
+				if (place === undefined) {
+					destination.places.push(itineraryItem.place!);
+				}
+			} else {
+				destinations.push({
+					destinationId: itineraryItemDestinationId,
+					places: [itineraryItem.place!],
+					mainMap: null,
+					secondaryMaps: []
+				});
+			}
+		});
+	});
+
+	return Promise.all(destinations.map(async (destination: PdfDestinationDataTmp): Promise<PdfDestination> => {
+		return {
+			destination: await getPlaceDetailed(destination.destinationId, '150x150'),
+			mainMap: destination.mainMap,
+			secondaryMaps: destination.secondaryMaps,
+			places: destination.places
+		};
+	}));
+}
+
+export function getFirstCity(place: Place): string {
+	if (isCity(place.id)) {
+		return place.id;
+	}
+	const reversedPlaceParentIds = place.parents.slice().reverse();
+
+	for (const parentId of reversedPlaceParentIds) {
+		if (isCity(parentId)) {
+			return parentId;
+		}
+	}
+	return reversedPlaceParentIds[0];
+}
+
+export function isCity(id: string): boolean {
+	const cities = ['city', 'town', 'village'];
+	const level = id.split(':')[0];
+	return level ? cities.includes(level) : false;
+}
+
+export async function getCollectionsForDestinations(destinations: PdfDestination[]): Promise<Collection[][]> {
```

Comment:
move to collection namespace accepts destination ids and returns Map.<id, Collections[]>